patch/QuadEnvelope/QuadEnvelope.cpp

```diff
@@ -0,0 +1,164 @@
+#include "daisy_patch.h"
+#include "daisysp.h"
+
+using namespace daisy;
+using namespace daisysp;
+
+DaisyPatch hw;
+
+int curveTimeMode;
+
+struct envStruct
+{
+    AdEnv env;
+    Parameter attackParam;
+    Parameter decayParam;
+    Parameter curveParam;
+    float envSig;
+};
+
+envStruct envelopes[2];
+void ProcessControls();
+
+void AudioCallback(float **in, float **out, size_t size)
+{
+    //Process control inputs
+    ProcessControls();
+    
+    for(size_t i = 0; i < size; i++)
+    {
+	//Get the next envelope samples
+	envelopes[0].envSig = envelopes[0].env.Process();
+	envelopes[1].envSig = envelopes[1].env.Process();
+		
+	for (size_t chn = 0; chn < 2; chn++)
+	{
+	    //The envelopes effect the outputs in pairs
+	    out[chn * 2][i] = in[chn * 2][i] * envelopes[chn].envSig;
+	    out[chn * 2 + 1][i] = in[chn * 2 + 1][i] * envelopes[chn].envSig;	    
+	}
+    }
+}
+
+void InitEnvelopes(float samplerate)
+{
+    for (int i = 0; i < 2; i++)
+    {
+	//envelope values and Init
+	envelopes[i].env.Init(samplerate);
+	envelopes[i].env.SetMax(1);
+	envelopes[i].env.SetMin(0);
+	envelopes[i].env.SetCurve(0);	
+    }
+
+    //envelope parameters (control inputs)
+    envelopes[0].attackParam.Init(hw.controls[0], .01, 2, Parameter::EXPONENTIAL);
+    envelopes[0].decayParam.Init(hw.controls[1], .01, 2, Parameter::EXPONENTIAL);
+    envelopes[0].curveParam.Init(hw.controls[0], -10, 10, Parameter::LINEAR);
+
+    envelopes[1].attackParam.Init(hw.controls[2], .01, 2, Parameter::EXPONENTIAL);
+    envelopes[1].decayParam.Init(hw.controls[3], .01, 2, Parameter::EXPONENTIAL);
+    envelopes[1].curveParam.Init(hw.controls[2], -10, 10, Parameter::LINEAR);
+}
+
+void UpdateOled();
+
+int main(void)
+{
+    float samplerate;
+    hw.Init();
+    samplerate = hw.AudioSampleRate();
+
+    InitEnvelopes(samplerate);
+
+    curveTimeMode = 0;
+    
+    UpdateOled();
+	
+    // Start the ADC and Audio Peripherals on the Hardware
+    hw.StartAdc();
+    hw.StartAudio(AudioCallback);
+    for(;;)
+    {
+	//Send the latest envelope values to the CV outs
+	dsy_dac_write(DSY_DAC_CHN1, envelopes[0].envSig * 4095);
+	dsy_dac_write(DSY_DAC_CHN2, envelopes[1].envSig * 4095);
+	hw.DelayMs(1);
+    }
+}
+
+void UpdateOled()
+{
+    hw.display.Fill(false);
+    
+    hw.display.SetCursor(0,0);
+    hw.display.WriteString("env1", Font_7x10, true);
+    
+    hw.display.SetCursor(70,0);
+    hw.display.WriteString("env2", Font_7x10, true);
+    
+    hw.display.SetCursor(0,50);
+
+    //curve or attack/decay mode
+    if (curveTimeMode)
+    {
+	hw.display.WriteString("curve", Font_7x10, true);
+    }
+    else
+    {
+	hw.display.WriteString("attack/decay", Font_7x10, true); 
+    }
```

Comment:
We may want to update the OLED driver to work with cpp strings, but for now since they're technically char* buffers, this throws warnings.

consider instead:
```C++
char buff[64]; // buffer of some size (can be reused for entire function)
sprintf(buff, "MyString -- And even a value: %d\n", some_value);
hw.display.WriteString(buff, Font_7x10, true);
```
```